tools/test/donkeyrider.py

```diff
@@ -0,0 +1,133 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import os
+import shutil
+import signal
+import sys
+
+stop = False
+
+
+def shutdown(signum, frame):
+    global stop
+    stop = True
+    print("Caught", signal.Signals(signum).name + ". Stopping..", file=sys.stderr)
+    sys.exit(1)
+
+
+def main():
+    if 'USE_PODMAN' not in os.environ or os.environ['USE_PODMAN'] == "":
+        docker_executable = shutil.which("docker")
+        podman_executable = shutil.which("podman")
+        if docker_executable and podman_executable:
+            with open(docker_executable, 'br') as fhandle:
+                if podman_executable.encode() in fhandle.read():
+                    os.environ['USE_PODMAN'] = "1"
+        elif docker_executable:
+            os.environ['USE_PODMAN'] = "0"
+        elif podman_executable:
+            os.environ['USE_PODMAN'] = "1"
+        else:
+            print("No compatible container executable (podman/docker) found! Exiting..", file=sys.stderr)
+            sys.exit(1)
+
+    project_dir = os.path.abspath(os.path.join(os.path.abspath(os.path.dirname(__file__)), "../.."))
+    print("Detected project directory", project_dir, file=sys.stderr)
+
+    parser = argparse.ArgumentParser(description='Run the full autotest pipeline.')
+    parser.add_argument('test_matrix_file', metavar='test matrix file', type=str, nargs='?', default=None,
+                        help='path to the test matrix')
+    parser.add_argument('--build', dest='build', action='store_true',
+                        help='build images, instead of only downloading')
+    parser.add_argument('-f', '--filter', dest='filter_tests', type=str, action='append',
+                        help='filter tests')
+    parser.add_argument('-d', '--dist', dest='dist', type=str, help='select distribution')
+    parser.add_argument('--py', '--python', dest='python', type=str, help='select Python version')
+    parser.add_argument('-s', '--suite', dest='suite', type=str, help='select test suite')
+    parser.add_argument('--db', '--database', dest='database', type=str, help='select RDBMS')
+    args = parser.parse_args()
+    if args.test_matrix_file is None:
+        matrix_file = os.path.join(project_dir, "etc/docker/test/matrix.yml")
+    else:
+        matrix_file = args.test_matrix_file
+    print("Using test matrix from file", matrix_file, file=sys.stderr)
+
+    # make sure the following imports work
```

Comment:
Explaining comments have some flaws: They most likely get forgotten when the code is refactored and indicate, that the code is not self explaining.

How about creating functions instead of the comments and blocks under them:
```python
def main():
    check_imports()
    args = parse_args()
    matrix = parse_martrix(args)
    images = build_images(matrix)
    run_tests(matrix, images, args.filter_tests)
```

Now the code clearly communicates the steps it takes, and you don't have to look into the `parse_args` code if `parse_martrix` has a problem.

Reply:
will have another look at this part of the code. I get your point.

Reply:
I have boiled down the main method to something more understandable. Of course it doesn't come close to your example, since there are more variables created and used in different parts of the code.